### PR TITLE
Make Context to be Device-agnostic Step by Step (1/N)

### PR DESCRIPTION
**Pull request by fffrog**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #137580
* #137578
* #136526
* __->__ #136519

----

- make init to be device-agnostic and move it to AcceleratorHooksInterface
- refactoring context related to device initialization

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

**Comment by FFFrog** [comment]
@ezyang Thanks for your approval, can I ignore the ci failure of xpu and merge the commit into truck? I will submit a new PR in another [repo](https://github.com/intel/torch-xpu-ops) to fix xpu ci later

**Comment by ezyang** [comment]
I don't know what xpu greenness policy is. But when we did analogous things for xla, we simultaneously updated the third party pin so that pytorch was always kept green. I suggest doing that too...

**Comment by FFFrog** [comment]
> I don't know what xpu greenness policy is. But when we did analogous things for xla, we simultaneously updated the third party pin so that pytorch was always kept green. I suggest doing that too...

I got it, thank you.

**Comment by FFFrog** [comment]
@EikanWang, @jgong5, sorry to bother you, could you please take a look for me? Handling xpu ci workflow failure is a little tricky for me.

**Review by EikanWang — APPROVED** [review]
LGTM

**Comment by FFFrog** [comment]
> @FFFrog , thanks for the PR submission. We are running the CI. But it seems like the torch-xpu-ops will be broken due to the missed `lazyInitDevice`. We will have an ad-hoc validation to check if there are any failures.

Thank you for your comments.

@EikanWang  @guangyey , what I want to add is that do I need to update [pin commit](https://github.com/pytorch/pytorch/blob/main/third_party/xpu.txt) for xpu when this [pr](https://github.com/intel/torch-xpu-ops/pull/934) is merge later? Or there is some mechanism that can be done automatically.

**Comment by guangyey** [comment]
> @EikanWang @guangyey , what I want to add is that do I need to update [pin commit](https://github.com/pytorch/pytorch/blob/main/third_party/xpu.txt) for xpu when this [pr](https://github.com/intel/torch-xpu-ops/pull/934) is merge later? Or there is some mechanism that can be done automatically.

Please help update xpu.txt directly once this [pr](https://github.com/intel/torch-xpu-ops/pull/934) is merged. Thanks for your help.

**Comment by FFFrog** [comment]
> > @EikanWang @guangyey , what I want to add is that do I need to update [pin commit](https://github.com/pytorch/pytorch/blob/main/third_party/xpu.txt) for xpu when this [pr](https://github.com/intel/torch-xpu-ops/pull/934) is merge later? Or there is some mechanism that can be done automatically.
> 
> Please help update xpu.txt directly once this [pr](https://github.com/intel/torch-xpu-ops/pull/934) is merged. Thanks for your help.

Thank you :D

**Review by guangyey — APPROVED** [review]

**Comment by etaf** [comment]
@EikanWang  @FFFrog  The following failed jobs have not been resolved but PR merged, which break current XPU CI.  I'll resolve this error in PR #137585. But suggest pass XPU CI before merge PR to save triage time.
[xpu / linux-jammy-xpu-py3.9 / test (default, 3, 4, linux.idc.xpu)](https://hud.pytorch.org/pr/pytorch/pytorch/136519#31209516897) ([gh](https://github.com/pytorch/pytorch/actions/runs/11226861852/job/31209516897))
inductor/test_torchinductor_dynamic_shapes.py::DynamicShapesGPUTests::test_scatter_add3_dynamic_shapes_xpu
[xpu / linux-jammy-xpu-py3.9 / test (default, 4, 4, linux.idc.xpu)](https://hud.pytorch.org/pr/pytorch/pytorch/136519#31209517023) ([gh](https://github.com/pytorch/pytorch/actions/runs/11226861852/job/31209517023))
inductor/test_torchinductor.py::GPUTests::test_scatter_add3_xpu

**Comment by jovianjaison** [comment]
@pytorchbot revert -m "this pr is causing errors internally" -c ghfirst
error: class "at::Context" has no member "lazyInitCUDA"

**Comment by FFFrog** [comment]
> #137585

Sorry for the late reply.
I will update this pr once [the pr](https://github.com/pytorch/pytorch/pull/137585) is merged.

File: third_party/xpu.txt
Comment (EikanWang) [inline third_party/xpu.txt:1]: We will update torch-xpu-ops commit pin first and then you can drop this change. @FFFrog

Comment (FFFrog) [reply third_party/xpu.txt:1]: Ok, have updated.

**Comment by ezyang** [comment]
@FFFrog so specifically the problem is there is some Caffe2 code internally that calls lazyInitCUDA. There are a few ways to fix this but I think adding aliases for the old method names is the easiest.

**Comment by FFFrog** [comment]
> @FFFrog so specifically the problem is there is some Caffe2 code internally that calls lazyInitCUDA. There are a few ways to fix this but I think adding aliases for the old method names is the easiest.

Thank you, I have updated the pr, please refer to the latest [commit](https://github.com/pytorch/pytorch/pull/136519/commits/cd098f38cc72662993a424f5df91a8a3486e05c9) to check what changed :D

**Comment by ezyang** [comment]
I'm going to help, it's just one small problem now

File: aten/src/ATen/detail/MTIAHooksInterface.h [hunk 1]
```diff
@@ -109,6 +109,11 @@ struct TORCH_API MTIAHooksInterface : AcceleratorHooksInterface {
     FAIL_MTIAHOOKS_FUNC(__func__);
     return nullptr;
   }
+
+  // Perserved for BC
+  virtual void initMTIA() const {
+    return;
+  }
```
Comment (ezyang) [inline aten/src/ATen/detail/MTIAHooksInterface.h:116]: I don't think this should exist

**Comment by ezyang** [comment]
I'm going to attempt to fix this directly, but a way to fix this from first principles in OSS, is to *not* rename any of the virtual methods on the hooks classes, and have a new implementation of init() in the subclass interfaces we let people subclass from that call the old method by default. What is done in this PR is cleaner but it's BC breaking and concretely what I have to do is rename initMTIA to init internally.

**Comment by FFFrog** [comment]
My principle is to make Context.h and DeviceInterface related files clearer and easier to understand, but the changes in the current pr do introduce BC breaking, which should not be the case for a widely used project like PyTorch.

> a way to fix this from first principles in OSS, is to _not_ rename any of the virtual methods on the hooks classes

I agree with your point of view, so I updated the PR and made the following changes:
- For the in-tree backend, the implementation of device initialization is moved to the newly added **init()**, and the call of the old interface **initCUDA()** will be forwarded to init, while ensuring the code logic is as clear and usable as possible, and BC is also guaranteed
- For the out-of-tree backend, in order to ensure BC as much as possible, no changes are made to **initMTIA()**, but it is forwarded to **initMITA()** in the new interface **init()**, and a new deprecation prompt message is added

**Comment by ezyang** [comment]
It turns out that the BC fix is pretty easy. I'm going to land my own copy as I need to also land an fbcode change

**Comment by ezyang** [comment]
Internal diff with the BC update fixes https://www.internalfb.com/diff/D64471142

**Comment by FFFrog** [comment]
@ezyang I will close this PR because the [PR](https://github.com/pytorch/pytorch/pull/138155) has been merged.

**Pull request closed without merging**